File: stepik-union/src/main/java/org/stepik/core/ui/Elements.java
package org.stepik.core.ui;

import org.jetbrains.annotations.NotNull;
import org.w3c.dom.Node;
import org.w3c.dom.html.HTMLCollection;
import org.w3c.dom.html.HTMLInputElement;
import org.w3c.dom.html.HTMLTextAreaElement;

import java.util.Iterator;

class Elements implements Iterable<Node> {
    private final HTMLCollection elements;

    Elements(@NotNull HTMLCollection elements) {
        this.elements = elements;
    }

    @NotNull
    String getAction() {
        Node item = elements.namedItem("action");
        if (item instanceof HTMLInputElement) {
            return ((HTMLInputElement) item).getValue();
        }

        return "";
    }

    @NotNull
    String getType() {
        Node item = elements.namedItem("type");
        if (item instanceof HTMLInputElement) {
            return ((HTMLInputElement) item).getValue();
        }

        return "";
    }

    @NotNull
    String getInputValue(@NotNull String name) {
        Node item = elements.namedItem(name);
        String value = null;
        if (item instanceof HTMLInputElement) {
            value = ((HTMLInputElement) item).getValue();
        } else if (item instanceof HTMLTextAreaElement) {
            value = ((HTMLTextAreaElement) item).getValue();
        }
        return value != null ? value : "";
    }

    boolean isFromFile() {
        HTMLInputElement element = ((HTMLInputElement) elements.namedItem("isFromFile"));
        return element != null && "true".equals(element.getValue());
    }

    long getAttemptId() {
        String value = getInputValue("attemptId");
        try {
            return Long.parseLong(value);
        } catch (NumberFormatException e) {
            return 0L;
        }
    }

    Boolean isLocked() {
        return Boolean.valueOf(getInputValue("locked"));
    }

    @NotNull
    @Override
    public Iterator<Node> iterator() {
        return new Iterator<Node>() {
            private final int size = elements.getLength();
            private int index;

            @Override
            public boolean hasNext() {
                return index < size;
            }

            @NotNull
            @Override
            public Node next() {
                return elements.item(index++);
            }
        };
    }
}

File: stepik-core/src/main/kotlin/org/stepik/core/ui/Elements.kt
package org.stepik.core.ui

import org.w3c.dom.Node
import org.w3c.dom.html.HTMLCollection
import org.w3c.dom.html.HTMLInputElement
import org.w3c.dom.html.HTMLTextAreaElement

internal class Elements(private val elements: HTMLCollection) : Iterable<Node> {

    val action: String
        get() = (elements.namedItem("action") as? HTMLInputElement)?.value ?: ""

    val type: String
        get() = (elements.namedItem("type") as? HTMLInputElement)?.value ?: ""

    val isFromFile: Boolean
        get() = (elements.namedItem("isFromFile") as? HTMLInputElement)?.value == "true"

    val attemptId: Long
        get() = getInputValue("attemptId").toLongOrNull() ?: 0L

    val isLocked: Boolean
        get() = getInputValue("locked").toBoolean()

    fun getInputValue(name: String): String {
        val item = elements.namedItem(name)
        return when (item) {
            is HTMLInputElement -> item.value
            is HTMLTextAreaElement -> item.value
            else -> ""
        }
    }

    override fun iterator(): Iterator<Node> {
        return object : Iterator<Node> {
            private val size = elements.length
            private var index: Int = 0

            override fun hasNext(): Boolean {
                return index < size
            }

            override fun next(): Node {
                return elements.item(index++)
            }
        }
    }
}
